### PR TITLE
Do not allow code with warnings to pass CI

### PR DESCRIPTION
**Pull request by npaun**
This PR enables `-Werror`, when building in the CI. As a result, warnings are treated as a fatal error. This ensures that warnings will be addressed, and won't pile up over time in our codebase. Here are some specific details about the configuration.

* Warnings aren't fatal locally. This is for convenience while working on code.
* Warnings aren't fatal in `external/`. This is because these warnings come from our dependencies, and need to be fixed there.
* Deprecation warnings and `#warning` are not fatal. They're just here to give us a heads up that something will need to be fixed eventually.
* Warnings are only fatal on Linux builds. I didn't want to make this a maddening quest of fixing macOS, Windows AND Linux compiler warnings every time we do a PR.

### CODE REVIEW

File: .github/workflows/test.yml [hunk 1]
```diff
@@ -84,6 +84,9 @@ jobs:
           sudo apt-get install -y --no-install-recommends clang-16 lld-16 libunwind-16 libc++abi1-16 libc++1-16 libc++-16-dev libclang-rt-16-dev
           echo "build:linux --action_env=CC=/usr/lib/llvm-16/bin/clang --action_env=CXX=/usr/lib/llvm-16/bin/clang++" >> .bazelrc
           echo "build:linux --host_action_env=CC=/usr/lib/llvm-16/bin/clang --host_action_env=CXX=/usr/lib/llvm-16/bin/clang++" >> .bazelrc
+          echo "build:linux --copt='-Werror'" >> .bazelrc
+          echo "build:linux --copt='-Wno-error=#warnings'" >> .bazelrc
+          echo "build:linux --copt='-Wno-error=deprecated-declarations'" >> .bazelrc
```
Comment (danlapid) [inline .github/workflows/test.yml:89]: Do we want need this if we have .bazelrc:47 `build --per_file_copt='external/v8@-Wno-deprecated-declarations'`
It's possible both are required, just not sure.

Comment (npaun) [reply .github/workflows/test.yml:89]: This would handle a project other than v8 having deprecated declarations.

Comment (danlapid) [reply .github/workflows/test.yml:89]: Yeah that's why I asked if we want this in any project other than v8.
I guess it's a more fitting question for @jasnell  maybe?

File: src/workerd/io/worker.c++ [hunk 2]
```diff
@@ -1602,7 +1602,8 @@ Worker::Worker(kj::Own<const Script> scriptParam,
                 }
                 KJ_CASE_ONEOF(limitError, kj::Exception) { }
               }
-            }
+            } else {}
```
Comment (anonrig) [inline src/workerd/io/worker.c++:1605]: These change seems unrelated. Can you remove it?

**Review by anonrig — COMMENTED** [review]
Can you also update `compile_flags.txt`?

**Comment by danlapid** [comment]
> Can you also update `compile_flags.txt`?

I think that would break "Warnings aren't fatal locally. This is for convenience while working on code."

File: src/workerd/io/worker.c++ (hunk 2 above)
Comment (npaun) [reply src/workerd/io/worker.c++:1605]: This fixes the one fatal warning we have currently

File: .github/workflows/test.yml (hunk 1 above)
Comment (fhanau) [reply .github/workflows/test.yml:89]: I think we regularly have deprecation warnings for JSG calling deprecated V8 functions (including right now?), it will be difficult to always have those under control.

File: .bazelrc [hunk 3]
```diff
@@ -43,6 +43,7 @@ build --@dawn//src/tint:tint_build_wgsl_writer=True
 # Our dependencies (ICU, zlib, etc.) produce a lot of these warnings, so we disable them. Depending
 # on the clang version, zlib either produces warnings for -Wdeprecated-non-prototype or does not
 # have that option, so disable -Wunknown-warning-option there too.
+build --per_file_copt='external/@-Wno-error'
```
Comment (fhanau) [inline .bazelrc:46]: nit: `external@-Wno-error` looks much cleaner to me

File: .github/workflows/test.yml (hunk 1 above)
Comment (danlapid) [reply .github/workflows/test.yml:89]: But these would fall under .bazelrc:47 no?

**Comment by fhanau** [comment]
> - Warnings aren't fatal locally. This is for convenience while working on code.
> - Warnings are only fatal on Linux builds. I didn't want to make this a maddening quest of fixing macOS, Windows AND Linux compiler warnings every time we do a PR.
> 

I think these two aspects are key, thank you for covering them. 1) is important for development velocity, 2) means we won't have to fix the many unhelpful Windows warnings we have. Including macOS via `build:unix` could be a good option as both systems share the vast majority of warnings when they happen but it conceivably could make things harder for non-mac developers and is not needed at this stage.

File: src/workerd/io/worker.c++ (hunk 2 above)
Comment (fhanau) [reply src/workerd/io/worker.c++:1605]: What is the warning? The statement looks really superfluous to me, if there's no other way to handle this let's add a comment here

File: .github/workflows/test.yml (hunk 1 above)
Comment (fhanau) [reply .github/workflows/test.yml:89]: No, that directive only adds the `-Wno-deprecated-declarations` flag when compiling source files whose paths 
match the `external/v8` pattern, i.e. V8 source files only.

File: src/workerd/io/worker.c++ (hunk 2 above)
Comment (npaun) [reply src/workerd/io/worker.c++:1605]: -Wdangling-else on KJ_IF_SOME. I see that James fixed a bunch of these in the past.  The macro tries to block that warning but it's still coming through. Maybe we should look into that.


Thinking out loud a bit: 

The macro's basic structure is 

```c++
if (has value)
   if (init; false) { 
      // no op
   } else 
```

So a usage expands to something like

```c++
if (has value) 
    if (init; false) { 
       // no op
    } else {
       // here is my actual block when there is a value 
    }
```

Clang is not happy about that outer if which has no braces.

Comment (npaun) [reply src/workerd/io/worker.c++:1605]: I can't trigger it on its own so I'm thinking either a clang bug, or something glitching with the pragmas

Comment (npaun) [reply src/workerd/io/worker.c++:1605]: I spent way too much time on this - the problem is triggered if KJ_IF_SOME is passed as an argument to any other macro.

e.g.
```c++
#define USELESS(x) x
USELESS(
    KJ_IF_SOME(foo, bar) {
       // blah
    }
)
```

Comment (npaun) [reply src/workerd/io/worker.c++]: For now I added a comment to explain why that else is there.

**Review by fhanau — APPROVED** [review]
✅

**Review by danlapid — APPROVED** [review]

**Review by danlapid — APPROVED** [review]
